### PR TITLE
Adds IRFFT Op to Signal Library

### PR DESCRIPTION
**Pull request by suleshahid**
Inverse-RFFT as part of Signal library ops.
Testing via current FFT Op tests.

BUG=[287346710](http://b/287346710)

### CODE REVIEW

File: signal/src/irfft_float.cc [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+size_t IrfftFloatGetNeededMemory(int32_t fft_length) {
+  size_t cfg_size = 0;
+  kiss_fft_float::kiss_fftr_alloc(fft_length, 1, nullptr, &cfg_size);
+  return sizeof(IrfftFloatState) + cfg_size;
+}
+
+void* IrfftFloatInit(int32_t fft_length, void* state, size_t state_size) {
+  IrfftFloatState* irfft_float_state = (IrfftFloatState*)state;
+  irfft_float_state->cfg =
+      (kiss_fft_float::kiss_fftr_cfg)(irfft_float_state + 1);
```
Comment (rascani) [inline signal/src/irfft_float.cc:41]: And this one to a reinterpret_cast

File: signal/src/irfft_float.cc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  irfft_float_state->fft_length = fft_length;
+  size_t cfg_size = state_size - sizeof(IrfftFloatState);
+  return kiss_fft_float::kiss_fftr_alloc(fft_length, 1, irfft_float_state->cfg,
+                                         &cfg_size);
+}
+
+void IrfftFloatApply(void* state, const Complex<float>* input, float* output) {
+  IrfftFloatState* irfft_float_state = (IrfftFloatState*)state;
```
Comment (rascani) [inline signal/src/irfft_float.cc:49]: And this to a static_cast

File: signal/micro/kernels/irfft.cc [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+template <typename T, size_t (*get_needed_memory_func)(int32_t),
+          void* (*init_func)(int32_t, void*, size_t)>
+void* Init(TfLiteContext* context, const char* buffer, size_t length) {
+  TFLITE_DCHECK(context->AllocatePersistentBuffer != nullptr);
+
+  const uint8_t* buffer_t = reinterpret_cast<const uint8_t*>(buffer);
+
+  auto* params = static_cast<TfLiteAudioFrontendIrfftParams*>(
+      context->AllocatePersistentBuffer(
+          context, sizeof(TfLiteAudioFrontendIrfftParams)));
```
[the remark this answers is not in the document]
Comment (suleshahid) [reply signal/micro/kernels/irfft.cc:60]: Done. Is there any preference over this vs TFLITE_DCHECK()?

File: signal/micro/kernels/irfft.cc [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+  TfLiteType fft_type;
+  int8_t* state;
+};
+
+template <typename T, size_t (*get_needed_memory_func)(int32_t),
+          void* (*init_func)(int32_t, void*, size_t)>
+void* Init(TfLiteContext* context, const char* buffer, size_t length) {
+  TFLITE_DCHECK(context->AllocatePersistentBuffer != nullptr);
+
+  const uint8_t* buffer_t = reinterpret_cast<const uint8_t*>(buffer);
```
[the remark this answers is not in the document]
Comment (suleshahid) [reply signal/micro/kernels/irfft.cc:56]: Done.

File: signal/micro/kernels/irfft.cc [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+      context->AllocatePersistentBuffer(
+          context, sizeof(TfLiteAudioFrontendIrfftParams)));
+
+  tflite::FlexbufferWrapper fbw(buffer_t, length);
+  params->fft_length = fbw.ElementAsInt32(kFftLengthIndex);
+  params->fft_type = typeToTfLiteType<T>();
+
+  size_t state_size = (*get_needed_memory_func)(params->fft_length);
+  params->state = reinterpret_cast<int8_t*>(
+      context->AllocatePersistentBuffer(context, state_size * sizeof(int8_t)));
```
[the remark this answers is not in the document]
Comment (suleshahid) [reply signal/micro/kernels/irfft.cc:68]: Done.

File: signal/src/irfft_float.cc [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+};
+
+size_t IrfftFloatGetNeededMemory(int32_t fft_length) {
+  size_t cfg_size = 0;
+  kiss_fft_float::kiss_fftr_alloc(fft_length, 1, nullptr, &cfg_size);
+  return sizeof(IrfftFloatState) + cfg_size;
+}
+
+void* IrfftFloatInit(int32_t fft_length, void* state, size_t state_size) {
+  IrfftFloatState* irfft_float_state = (IrfftFloatState*)state;
```
[the remark this answers is not in the document]
Comment (suleshahid) [reply signal/src/irfft_float.cc:39]: Done.

File: signal/src/irfft_float.cc (hunk 1 above)
Comment (suleshahid) [reply signal/src/irfft_float.cc:41]: Done.

File: signal/src/irfft_float.cc (hunk 2 above)
Comment (suleshahid) [reply signal/src/irfft_float.cc:49]: Done.

**Comment by suleshahid** [comment]
Ready for another look.

**Review by rascani — APPROVED** [review]

**Pull request merged**